### PR TITLE
Start extracting ast node type specification from node construction

### PR DESCRIPTION
**Pull request by pcardune**
I'm very excited about this pull request.

The goal here is to start separating the code that defines the types of nodes in a language's abstract syntax tree from the construction of those nodes, and eventually from the rendering/pretty-printing/describing of those nodes.

The idea is that we have some way of defining the shape of the AST. i.e. The AST for language X has nodes of types A, B, and C. Nodes of type A have properties u, v, and w, while nodes of type B have properties i and k. Each of the properties must in turn be of a particular type: another node, or list of nodes of types A, B, or C, or literal value. And each of the properties have a particular order, used for navigation between nodes: i.e. starting from a node of type A, we navigate to it's child property u before its child property v, and so on.

Once we have the shape of the AST properly typed, then we can write functions over the AST that are also well typed. So we can write a function that renders nodes of type A to a react element, and it will have access to the field u, v, and w of that node. Same thing with a function for pretty printing a node or describing a node, or whatever.

And finally, with all the above, we can write a parser which constructs an AST containing nodes of the correct type. All the functions that operate on those nodes don't care about the parser, and the parser doesn't care about them, making them mutually interchangeable.  Likewise, those functions that operate on nodes of the AST are just regular pure functions and can be composed to get extensibility/shared behavior across functions for different node types. Also, the shape of the AST can be extended with additional node types and all the existing functions for operating on the existing node types can be reused.

We already have a mechanism for defining the shapes of individual nodes in an AST: nodeSpecs. This PR does two main things:

1. enhances node specs with additional type declarations that make it possible to bring the types they implicitly define into typescript.

   With this enhancement, we can declare a nodeSpec like usual:
   ```ts
   const functionDefSpec = nodeSpec([
     required("name"),
     list("params"),
     optional("body"),
   ]);
   ```
   But now we can say that the type of a variable must conform to that nodeSpec:
   ```ts
   let myFunctionDef: FieldsForSpec<typeof functionDefSpec>;

   // the following fails because params is missing
   myFunctionDef = {name: new ASTNode(...), body: null};

   // the following fails because params is supposed to be a list of ASTNodes, not a list of strings
   myFunctionDef = {name: new ASTNode(...), body: null, params: ["a","b"]};
   ```
   and typescript will enforce this conformance to the nodespec statically!

2. creates a way to combine nodespecs together into an overarching "ast spec", and from that generate strongly typed functions for constructing an AST that matches that spec.

   Now we have a `createASTSpec()` function that takes a mapping from node type strings to their nodeSpecs that can be used like this:

   ```ts
   const weschemeASTSpec = createASTSpec({
     functionDef: nodeSpec(...),
     functionApp: nodeSpec(...),
     literal: nodeSpec(...),
     ...
   });
   ```

   Then using that language specific ast spec, we can construct nodes for that ast that are guaranteed to conform to the overall ast spec:

   ```ts
   weschemeASTSpec.makeNode({
      // typescript will ensure that this is one of the types specified
      // in the call to createASTSpec
      type: "functionDef",
      fields: {
        // typescript ensures that these fields match those defined
        // for "functionDef" node types
        name: ...,
        params: ...,
        body: ...,
      },
      pretty: (node) => {
        // typescript will ensure that the only fields you can access from node
        // are those defined by the spec for the "functionDef" node type.
        return `
          (defun ${node.fields.name}
            (${node.fields.params.join(' ')})
            (${node.fields.body})
          )`;
      },
      ...
   });
   ```

So the upshot of all this is that each language will be able to define a spec for its overall AST, and then you'll be able to write parsers/renderers/prettiers that generate or consume that AST and typescript will be able to guarantee that your parser only generates valid ASTs and that the renderers/prettiers/etc only consume the AST in a way that is typesafe. If all the components use typescript, then there is no need for any runtime validation of the AST because you simply can't write code that will break.

In my limited experience, getting this sort of static type safety requires you to compile the spec into a set of type declarations the same way you would compile a grammar into a parser with a parser generator. See [the spec](https://github.com/pcardune/snax/blob/main/snax/projects/snax/spec.ts), [spec compiler](https://github.com/pcardune/snax/blob/main/snax/projects/snax/spec-build.ts), and [compiled types](https://github.com/pcardune/snax/blob/main/snax/projects/snax/spec-gen.ts) I'm using for my toy language as an example. But what I've got in this PR lets us get type safety for any spec without an extra build step, while preserving the ability to do runtime checks.

### CODE REVIEW

File: packages/codemirror-blocks/src/nodes.tsx [hunk 1]
```diff
@@ -290,12 +357,13 @@ export function LambdaExpression(
     longDescription(node, level) {
       return `an anonymous function of ${pluralize(
         "argument",
-        node.fields.args.fields.ids
+        // TODO(pcardune): figure out how to describe node specs
+        // where a child must be a node of a particular type
+        node.fields.args.fields.ids as unknown[]
```
Comment (pcardune) [inline packages/codemirror-blocks/src/nodes.tsx:362]: So here's something interesting that popped up. Sometimes we have node types where the child nodes must be a particular type. nodeSpecs never did this deeper level of validation, and so its been possible to construct ASTs that validate but which blow up when they are rendered/described/prettied. Now the blowing up possibility is brought to light by typescript. I see two options:
1. Force functions that want to reach into a node's grand-children to do their own runtime validation or
2. provide a mechanism for enforcing this in the node spec.

For option 2, we then have to decide between nominal typing or structural typing. Do you say, "the args field must be a node of type identifierList" or do you say "the args field must be a node with an ids field which must be a list", or do you allow both. Then you also have the possibility of union types saying "the args field must be a node of one of the following types"

File: packages/wescheme-blocks/src/ast.tsx [hunk 2]
```diff
@@ -16,50 +17,79 @@ function pluralize(noun: string, set: unknown[]) {
...
-              P.standardSexpr(expr.fields.name, [expr.fields.body])
-            ),
+            node.fields.bindings.fields.exprs.map((expr) => {
+              if (!WeSchemeAST.isNodeOfType(expr, "variableDefinition")) {
+                throw new Error(
+                  `Expected letLikeExpr to have sequence of variableDefinition nodes`
+                );
+              }
+              return P.standardSexpr(expr.fields.name, [expr.fields.body]);
+            }),
```
Comment (pcardune) [inline packages/wescheme-blocks/src/ast.tsx:92]: Here is an example of doing runtime type checking to narrow down the type of grandchildren (and great grandchildren) in the pretty functions. This validation wasn't happening at all before. Now typescript forces us to do the validation at runtime because it doesn't have enough information to do it statically (it's not in the node spec!)

File: packages/wescheme-blocks/src/ast.tsx [hunk 3]
```diff
@@ -16,50 +17,79 @@ function pluralize(noun: string, set: unknown[]) {
   return set.length + " " + noun + (set.length != 1 ? "s" : "");
 }
 
+const WeSchemeAST = createASTSpec({
+  ...CoreAST.config,
+  letLikeExpr: {
+    spec: Spec.nodeSpec([
+      Spec.value<string, "form">("form"),
```
Comment (pcardune) [inline packages/wescheme-blocks/src/ast.tsx:24]: This part is also a little bit weird. Our code expects value fields to be a certain type, which was never actually defined in the spec. With typescript, we can specify it for typescript code (it doesn't turn into a runtime check) using a generic, but then we also have to specify the field name in the type.

File: packages/codemirror-blocks/src/nodes.tsx (hunk 1 above)
Comment (schanzer) [reply packages/codemirror-blocks/src/nodes.tsx]: I'm inclined to sat nominal typing. There are plenty of language constructs that might have identical fields but mean semantically different things. For example,  variable assignment and variable bindings, will both have `name` and `value` fields, but only one is allowed inside a function declaration.

Union types are intriguing. At that point, the nodeSpecs could almost be generated directly from a language's grammar.

File: packages/wescheme-blocks/src/ast.tsx (hunk 3 above)
Comment (schanzer) [reply packages/wescheme-blocks/src/ast.tsx:24]: Do you see this as a problem, for now? I agree that it's non-uniform compared to other specs, but I want us to leave the door open for languages to export a `typecheck : AST -> AST` function that stores type information in nodes that can be used by the `render()` function.

**Comment by schanzer** [comment]
Agreed - this is _really_ exciting! Two quick questions:

1) I'm assuming that, in your example, one could also define a `render` function in the object passed to `weschemeAST.makeNode()`? And that function would have to return some kind of vDOM-ish object that React can render?
2) Are there default `pretty` and `render` functions for nodes? Can they be overridden? Can someone override a default function for `weschemeAST` by using `makeNode` and passing a `functionDef` with a custom renderer? (Not asking because these are things I want -- asking because I want to understand the boundaries of what this checking/decoupling allows)

**Comment by pcardune** [comment]
Yes, all this should be possible.

**Defaults**

The more I think about this, the more layers I discover to the problem. Fundamentally we are trying to associate behavior (some method/function) with specific nodes in the AST, and make an easy to use API for configuring that association that handles the common cases with minimal code.

Defaults with overrides would let you get something working very quickly, but still be able to customize as needed. I see the following layers of increasing specificity with respect to behavior association:

1. default behavior for a node type across all languages

   This is ideal for node types and behaviors that are very common across many languages. For example, a "comment" node could be rendered and described in the same way across nearly every language. Same with conditional branching.

2. default behavior for a node type within a specific language family

   There are many flavors of lisp/scheme, but they all use s-expressions as their fundamental building block, so it would make sense for there to be a default way to render/describe s-expressions.

3. default behavior for a node type within a specific language

   Certain flavors of scheme, like wescheme have very specific language constructs that don't exist in other flavors of scheme. It would make sense for there to be a default behavior for the nodes representing this construct.

4. non-default behavior for a specific node within the AST

   This is essentially an escape hatch out of the AST spec abstraction. If you want to associate a behavior with a specific node in the AST that is not shared with any other node in the AST (not even with other nodes of the same type), then you have to do the association during parsing when you construct the specific node.

   We support this today, but I'm not sure it's necessary, and I'm not sure it's a good idea to support even if we can. If you need the escape hatch, that indicates there is something wrong with our API/abstractions that should be fixed more holistically. It also makes for less predictable behavior with AST edits/patching that don't go through the parser.

But there is still more complexity ahead: not all behaviors can have sensible defaults at the same layer. Rendering behavior and describing behavior can live together in the same default layer, but pretty printing behavior often needs to be more specific.

For example, rendering and describing a "comment" node type could have defaults specified at layer 1 (it's the same across all languages), but pretty printing a "comment" would have to be specific to layer 2 or layer 3. A lot of languages use // for comments, but a lot of them don't, so if a language module wants to use comments, it needs to provide its own implementation for pretty printing comments.

This is the part that I'm currently stuck on. Where do these behavior associations live? In the AST spec itself or somewhere else? What's the best way to provide "partial" implementations? At what stage do we check that an implementation is complete: at runtime when attempting to execute the behavior for a given node, parse time when the ast is generated from a spec, or statically at compile time with typescript? It's the same problem that abstract classes and inheritance attempts to solve, but I don't want to solve it with abstract classes and inheritance...

**Comment by schanzer** [comment]
**1 and 2** : I think these behaviors living in the CMB package itself -- we should try to provide sensible default node types and renderers for common node types that show up in _all_ or _many_ languages. I don't think this list needs to be exhaustive for 1.0, and the only thing I could _imagine_ adding is some kind of looping construct. Note: I don't think pretty-printing should live here.

**3**: A language-specific node should live in the language package.

**4**: I am in firm agreement that we should not support this. It feels like a recipe for disaster.

You're spot-on about certain behaviors having to live in the higher levels. I think it's fair to say that a language MUST include pretty-printers for all of its node types, since those are both (1) language specific and (2) pretty-printing and parsing will often need to be tested together, so they might as well live together! My API design experience is basically zero., so I don't have a sense for whether these more sense in the AST spec or elsewhere. 

As for when we check -- I'd like to rule out doing this at parse time, if possible. Parsing (1) happens all the time, (2) is a potential UI blocker, and (3) behavioral completeness across all nodetypes doesn't change dynamically. I lean towards putting them in the AST spec, so that when a language is loaded we determine whether there are any undefined behaviors. I'd _love_ to have this happen via TypeScript, but wouldn't that mean we require language authors to use TypeScript as well?

**Comment by pcardune** [comment]
Marking this as ready for review because I don't think this PR should get any bigger and all the other things we discussed can be handled separately/later.

**Comment by schanzer** [comment]
I'm comfortable merging this. Would you mind updating from master first?

**Comment by pcardune** [comment]
ok, should be good to go now...
